### PR TITLE
import legacy/non-legacy zklogin accounts

### PR DESCRIPTION
**Pull request by tlmysten**
## Description 

Describe the changes or additions included in this PR.

## Test plan 

How did you test the new or updated feature?

---

## Release notes

Check each box that your changes affect. If none of the boxes relate to your changes, release notes aren't required.

For each box you select, include information after the relevant heading that describes the impact of your changes that a user might notice and any actions they must take to implement updates. 

- [ ] Protocol: 
- [ ] Nodes (Validators and Full nodes): 
- [ ] Indexer: 
- [ ] JSON-RPC: 
- [ ] GraphQL: 
- [ ] CLI: 
- [ ] Rust SDK:
- [ ] REST API:

### CODE REVIEW

File: apps/wallet/src/background/accounts/zklogin/ZkLoginAccount.ts [hunk 1]
```diff
@@ -91,6 +94,19 @@ export function isZkLoginAccountSerializedUI(
...
 }
 
+async function hasTransactionHistory(address: string): Promise<boolean> {
+	const activeNetwork = await networkEnv.getActiveNetwork();
+	if (activeNetwork.env !== API_ENV.mainnet) {
+		return false;
+	}
+
+	const rpc = getSuiClient(activeNetwork);
+	const txns = await fetchTransactionsByAddress(rpc, address);
```
Comment (Jordan-Mysten) [inline apps/wallet/src/background/accounts/zklogin/ZkLoginAccount.ts:104]: Honestly I wouldn't bother doing this generalization, it's just two queries, not worth doing this for.

File: apps/wallet/src/background/accounts/zklogin/ZkLoginAccount.ts [hunk 2]
```diff
@@ -91,6 +94,19 @@ export function isZkLoginAccountSerializedUI(
 	return account.type === 'zkLogin';
 }
 
+async function hasTransactionHistory(address: string): Promise<boolean> {
+	const activeNetwork = await networkEnv.getActiveNetwork();
+	if (activeNetwork.env !== API_ENV.mainnet) {
```
Comment (Jordan-Mysten) [inline apps/wallet/src/background/accounts/zklogin/ZkLoginAccount.ts:99]: You shouldn't need this check

File: apps/wallet/src/background/accounts/zklogin/ZkLoginAccount.ts [hunk 3]
```diff
@@ -91,6 +94,19 @@ export function isZkLoginAccountSerializedUI(
...
+		return false;
+	}
+
+	const rpc = getSuiClient(activeNetwork);
```
Comment (Jordan-Mysten) [inline apps/wallet/src/background/accounts/zklogin/ZkLoginAccount.ts:103]: Should only be mainnet

File: apps/wallet/src/background/accounts/zklogin/ZkLoginAccount.ts (hunk 1 above)
Comment (Jordan-Mysten) [reply apps/wallet/src/background/accounts/zklogin/ZkLoginAccount.ts:104]: This also will be weird with graphQL so I'd revert this generalization for sure

File: apps/wallet/src/background/accounts/zklogin/ZkLoginAccount.ts [hunk 4]
```diff
@@ -148,6 +175,25 @@ export class ZkLoginAccount
...
+		ret.push({
+			...accountData,
+			address: legacyAddress,
+		});
+		if (normalizeSuiAddress(legacyAddress) !== normalizeSuiAddress(nonLegacyAddress)) {
+			if (await hasTransactionHistory(nonLegacyAddress)) {
+				ret.push({
+					...accountData,
+					address: nonLegacyAddress,
+					nickname: null,
```
Comment (Jordan-Mysten) [inline apps/wallet/src/background/accounts/zklogin/ZkLoginAccount.ts:191]: Probably should still set this, but just with `(address 2)` appended per the discussion

File: apps/wallet/src/background/accounts/zklogin/ZkLoginAccount.ts [hunk 5]
```diff
@@ -91,6 +100,26 @@ export function isZkLoginAccountSerializedUI(
...
+			},
+		}),
+		rpc.queryTransactionBlocks({
+			filter: {
+				FromAddress: address!,
+			},
+		}),
+	]);
+
+	return !!txnIds.data.length && !!fromTxnIds.data.length;
```
Comment (hayes-mysten) [inline apps/wallet/src/background/accounts/zklogin/ZkLoginAccount.ts:118]:
```suggestion
	return !!txnIds.data.length || !!fromTxnIds.data.length;
```

Comment (tlmysten) [reply apps/wallet/src/background/accounts/zklogin/ZkLoginAccount.ts]: omg

File: apps/wallet/src/background/accounts/zklogin/ZkLoginAccount.ts [hunk 6]
```diff
@@ -91,6 +100,26 @@ export function isZkLoginAccountSerializedUI(
 	return account.type === 'zkLogin';
 }
 
+async function hasTransactionHistory(address: string): Promise<boolean> {
+	const rpc = await getActiveNetworkSuiClient();
+	const [txnIds, fromTxnIds] = await Promise.all([
+		rpc.queryTransactionBlocks({
+			filter: {
```
Comment (Jordan-Mysten) [inline apps/wallet/src/background/accounts/zklogin/ZkLoginAccount.ts:107]: Limit 1 should make this faster

File: apps/wallet/src/background/accounts/zklogin/ZkLoginAccount.ts [hunk 7]
```diff
@@ -91,6 +100,26 @@ export function isZkLoginAccountSerializedUI(
 	return account.type === 'zkLogin';
 }
 
+async function hasTransactionHistory(address: string): Promise<boolean> {
+	const rpc = await getActiveNetworkSuiClient();
```
Comment (Jordan-Mysten) [inline apps/wallet/src/background/accounts/zklogin/ZkLoginAccount.ts:104]: You should just use the mainnet client

File: apps/wallet/src/background/accounts/zklogin/ZkLoginAccount.ts [hunk 8]
```diff
@@ -282,12 +342,66 @@ export class ZkLoginAccount
 			randomness: randomness.toString(),
 			jwt,
 		};
-		ephemeralValue[serializeNetwork(activeNetwork)] = credentialsData;
-		await this.setEphemeralValue(ephemeralValue);
+		await this.saveCredentialsData(credentialsData);
```
Comment (tlmysten) [inline apps/wallet/src/background/accounts/zklogin/ZkLoginAccount.ts:345]: extracting to a method so we can update credentials for the other address if it exists

File: apps/wallet/src/background/accounts/zklogin/ZkLoginAccount.ts [hunk 9]
```diff
@@ -282,12 +342,63 @@ export class ZkLoginAccount
...
+		// complementary account has been imported before and is unlocked, we
+		// update its credentials. This sync does not need to block the login
+		// process.
+		this.#syncAlternateAccount(jwt, credentialsData);
+
 		return credentialsData;
 	}
 
+	async #syncAlternateAccount(jwt: string, credentialsData: CredentialData) {
+		const currentAddress = await this.address;
```
Comment (Jordan-Mysten) [inline apps/wallet/src/background/accounts/zklogin/ZkLoginAccount.ts:359]: This check seems really weird, why is it using the current address? You should be able to do this based on just the stored accounts not based on the current state

File: apps/wallet/src/background/accounts/zklogin/ZkLoginAccount.ts [hunk 10]
```diff
@@ -282,12 +342,63 @@ export class ZkLoginAccount
...
+				{
+					address,
+					createdAt: Date.now(),
+					publicKey: cached.publicKey,
+					selected: false,
+					type: 'zkLogin',
+					nickname: cached.nickname ? `${cached.nickname} (address 2)` : null,
+					lastUnlockedOn: null,
+				},
+			]);
```
Comment (hayes-mysten) [inline apps/wallet/src/background/accounts/zklogin/ZkLoginAccount.ts:392]: Do we need something like this here too?

```suggestion
			const [newAccount] = await addNewAccounts([
				{
					address,
					createdAt: Date.now(),
					publicKey: cached.publicKey,
					selected: false,
					type: 'zkLogin',
					nickname: cached.nickname ? `${cached.nickname} (address 2)` : null,
					lastUnlockedOn: null,
				},
			]);
			
			if (existing instanceof ZkLoginAccount && !(await existing.isLocked())) {
				existing.saveCredentialsData(_credentialsData);
			}
```

File: apps/wallet/src/background/accounts/zklogin/ZkLoginAccount.ts [hunk 11]
```diff
@@ -282,12 +342,63 @@ export class ZkLoginAccount
...
+			await this.#importOrUpdateAlternateAccount(nonLegacyAddress, credentialsData);
+		}
+	}
+
+	// TODO XXX FIXME: update alternate account credentials
+	async #importOrUpdateAlternateAccount(address: string, _credentialsData: CredentialData) {
+		const imported = await getAccountsByAddress(address);
+		if (imported.length) {
+			const [existing] = imported;
+			if (existing instanceof ZkLoginAccount && !(await existing.isLocked())) {
```
Comment (hayes-mysten) [inline apps/wallet/src/background/accounts/zklogin/ZkLoginAccount.ts:377]: Whats the expected behavior when unlocking?  I think if we are unlocking both accounts we don't need this check, but if we aren't unlocking the alternative accounts we probably don't need to save the credentials?

not 100% sure I know how these values are being used though

File: apps/wallet/src/background/accounts/zklogin/ZkLoginAccount.ts [hunk 12]
```diff
@@ -282,12 +342,63 @@ export class ZkLoginAccount
...
+		if (normalizeSuiAddress(legacyAddress) !== normalizeSuiAddress(currentAddress)) {
+			await this.#importOrUpdateAlternateAccount(legacyAddress, credentialsData);
+		}
+		if (normalizeSuiAddress(nonLegacyAddress) !== normalizeSuiAddress(currentAddress)) {
+			await this.#importOrUpdateAlternateAccount(nonLegacyAddress, credentialsData);
+		}
+	}
+
+	// TODO XXX FIXME: update alternate account credentials
+	async #importOrUpdateAlternateAccount(address: string, _credentialsData: CredentialData) {
```
Comment (hayes-mysten) [inline apps/wallet/src/background/accounts/zklogin/ZkLoginAccount.ts:373]: the todo and `_` can be removed now right?

**Review by hayes-mysten — APPROVED** [review]

**Review by mamos-mysten — APPROVED** [review]

**Pull request merged**